modules/images/image-loading-optimization/class-ilo-html-tag-processor.php

```diff
@@ -0,0 +1,301 @@
+<?php
+/**
+ * Image Loading Optimization: ILO_HTML_Tag_Processor class
+ *
+ * @package performance-lab
+ * @since n.e.x.t
+ */
+
+/**
+ * Processor leveraging WP_HTML_Tag_Processor which gathers breadcrumbs which can be queried while iterating the open_tags() generator .
+ *
+ * @since n.e.x.t
+ * @access private
+ */
+final class ILO_HTML_Tag_Processor {
+
+	/**
+	 * HTML void tags (i.e. those which are self-closing).
+	 *
+	 * @link https://html.spec.whatwg.org/multipage/syntax.html#void-elements
+	 * @see WP_HTML_Processor::is_void()
+	 * @todo Reuse `WP_HTML_Processor::is_void()` once WordPress 6.4 is the minimum-supported version.
+	 *
+	 * @var string[]
+	 */
+	const VOID_TAGS = array(
+		'AREA',
+		'BASE',
+		'BASEFONT', // Obsolete.
+		'BGSOUND', // Obsolete.
+		'BR',
+		'COL',
+		'EMBED',
+		'FRAME', // Deprecated.
+		'HR',
+		'IMG',
+		'INPUT',
+		'KEYGEN', // Obsolete.
+		'LINK',
+		'META',
+		'PARAM', // Deprecated.
+		'SOURCE',
+		'TRACK',
+		'WBR',
+	);
+
+	/**
+	 * The set of HTML tags whose presence will implicitly close a <p> element.
+	 * For example '<p>foo<h1>bar</h1>' should parse the same as '<p>foo</p><h1>bar</h1>'.
+	 *
+	 * @link https://html.spec.whatwg.org/multipage/grouping-content.html#the-p-element
+	 *
+	 * @var string[]
+	 */
+	const P_CLOSING_TAGS = array(
+		'ADDRESS',
+		'ARTICLE',
+		'ASIDE',
```

Comment:
Why do we need these lists here? What specifically in regards to void and p closing tags is needed here, and why? Is there a bug in the `WP_HTML_Tag_Processor`, or crucial limitation?

It would be great if you could clarify the need for all this in relation to this module. Right now I'm wondering whether this module should need all those functionalities - like why are those special tags relevant for image optimization?

Reply:
And yeah, these lists of tags are all related to image optimization in that we need to know them in order to accurately construct breadcrumbs for any given tag.

Reply:
Thanks, that makes sense. I wonder how this version of the tag processor is for performance.

For the plugin we don't need to worry about this too much, but for a potential core proposal we would need to spend more time on this - particularly as I'm not sure we would want two separate versions of the tag processor in core. And we will only be able to use this one throughout if it doesn't have notable negative performance implications.

Reply:
It should have about the same performance as `WP_HTML_Tag_Processor` as it is just iterating over tags and is doing much less than `WP_HTML_Processor`, which is less predictable for performance.